File: ChessServer.m
//
//  ChessServer.h
//  Kibitz
//
//  Copyright 2014 William Entriken, licensed under the MIT license:
//  http://opensource.org/licenses/MIT
//
//  Based on Kibitz / ChessServer 2006 Klaus Thul
//

#import "ChessServer.h"

@implementation ChessServer

- (NSString *)description
{
    return self.serverName;
}

- (instancetype)initWithCoder:(NSCoder *)coder
{
    self = [super init];
    if (self) {
        self.serverName = [coder decodeObjectForKey:@"serverName"];
        self.serverAddress = [coder decodeObjectForKey:@"serverAddress"];
        self.serverPort = [coder decodeObjectForKey:@"serverPort"];
        self.userName = [coder decodeObjectForKey:@"userName"];
        self.userPassword = [coder decodeObjectForKey:@"userPassword"];
        self.initCommands = [coder decodeObjectForKey:@"initCommands"];
        self.useTimeseal = [coder decodeBoolForKey:@"useTimeseal"];
        self.connectAtStartup = [coder decodeBoolForKey:@"connectAtStartup"];
        self.issueSeek = [coder decodeBoolForKey:@"issueSeek"];
        self.seek = [coder decodeObjectForKey:@"seek"];
    }
    return self;
}

- (void)encodeWithCoder:(NSCoder *)coder
{
	[coder encodeObject: self.serverName forKey: @"serverName"];
	[coder encodeObject: self.serverAddress forKey: @"serverAddress"];
	[coder encodeObject: self.serverPort forKey: @"serverPort"];
	[coder encodeObject: self.userName forKey: @"userName"];
	[coder encodeObject: self.userPassword forKey: @"userPassword"];
	[coder encodeObject: self.initCommands forKey: @"initCommands"];
	[coder encodeBool: self.useTimeseal forKey: @"useTimeseal"];
	[coder encodeBool: self.connectAtStartup forKey: @"connectAtStartup"];
	[coder encodeBool: self.issueSeek forKey: @"issueSeek"];
	[coder encodeObject: self.seek forKey: @"seek"];
}

@end

File: ChessServer.swift
//
//  ChessServer.swift
//  Kibitz
//
//  Copyright 2014 William Entriken, licensed under the MIT license:
//  http://opensource.org/licenses/MIT
//
//  Based on Kibitz / ChessServer 2006 Klaus Thul
//

@objc(ChessServer) // http://stackoverflow.com/a/24329118/300224
class ChessServer: NSObject, NSCoding {
    var serverName: String = "Free Internet Chess Server (FICS)"
    var serverAddress: String = "freechess.org"
    var serverPort: Int = 53 // another popular option is 5000
    var userName: String = "guest"
    var userPassword: String = "guest"
    var initCommands: String = " "
    var useTimeseal: Bool = false
    var connectAtStartup: Bool = false
    var issueSeek: Bool = false
    var seek: Seek = Seek()
    
    override var description: String {
        return self.serverName
    }
    
    override required init() {
        super.init()
    }

    required convenience init(coder: NSCoder) {
        self.init()
        
        if let serverName = coder.decodeObject(forKey: "serverName") as? String {
            self.serverName = serverName
        }
        if let serverAddress = coder.decodeObject(forKey: "serverAddress") as? String {
            self.serverAddress = serverAddress
        }
        if let serverPort = coder.decodeObject(forKey: "serverPort") as? Int {
            self.serverPort = serverPort
        }
        if let userName = coder.decodeObject(forKey: "userName") as? String {
            self.userName = userName
        }
        if let userPassword = coder.decodeObject(forKey: "userPassword") as? String {
            self.userPassword = userPassword
        }
        if let initCommands = coder.decodeObject(forKey: "initCommands") as? String {
            self.initCommands = initCommands
        }
        self.useTimeseal = coder.decodeBool(forKey: "useTimeseal")
        self.connectAtStartup = coder.decodeBool(forKey: "connectAtStartup")
        self.issueSeek = coder.decodeBool(forKey: "issueSeek")
        if let seek = coder.decodeObject(forKey: "seek") as? Seek {
            self.seek = seek
        }
    }
    
    func encode(with aCoder: NSCoder) {
        aCoder.encode(self.serverName, forKey: "serverName")
        aCoder.encode(self.serverAddress, forKey: "serverAddress")
        aCoder.encode(self.serverPort, forKey: "serverPort")
        aCoder.encode(self.userName, forKey: "userName")
        aCoder.encode(self.userPassword, forKey: "userPassword")
        aCoder.encode(self.initCommands, forKey: "initCommands")
        aCoder.encode(self.useTimeseal, forKey: "useTimeseal")
        aCoder.encode(self.connectAtStartup, forKey: "connectAtStartup")
        aCoder.encode(self.issueSeek, forKey: "issueSeek")
        aCoder.encode(self.seek, forKey: "seek")
    }
}
